### PR TITLE
Systemview and RTT features for debugging firmware on device

### PR DESCRIPTION
**Pull request by hiviah**
When built with `SYSTEM_VIEW=1` (e.g. `SYSTEM_VIEW=1 make build_firmware`):

  * profiles interrupt timings, currently SysTick and USB_HS, others can be added
  * micropython's `print` and C `printf`output appears in Systemview log instead of USB serial port
  * destination of messages can be switched to RTT (real-time terminal) by setting `SYSTEMVIEW_DEST_SYSTEMVIEW  (0)` in `mpconfigport.h` - RTT is supported also in `openocd`
  * message transfers are basically just memcpy, so are much faster (100+ times) and have higher throughput than USB serial
  * logging works in full firmware (BITCOIN_ONLY=0) where we normally don't have serial port
  * you can add your own logging for other values, other interrupts, useful in case of race conditions

Note: RTT output option only supports "terminal-like" messages including color, but not the interrupt timings.

Note: the Segger files added are under 1-clause BSD license.

Depends on this PR for the print redirection in uPy: https://github.com/trezor/micropython/pull/5

uPy's `print` is redirected by changing `MP_PLAT_PRINT_STRN` in `mpconfigport.h`.

### CODE REVIEW

File: core/SConscript.firmware [hunk 1]
```diff
@@ -349,6 +350,22 @@ if RDI:
...
+        'embed/segger/SEGGER/SEGGER_RTT_ASM_ARMv7M.S',
+        'embed/segger/SEGGER/Syscalls/SEGGER_RTT_Syscalls_GCC.c',
+        'embed/firmware/systemview.c',
+    ]
+    CPPPATH_MOD += [
+        'embed/segger/SEGGER/',
+        'embed/segger/Config/',
+    ]
+    CPPDEFINES_MOD += ['SYSTEM_VIEW']
+    CCFLAGS_MOD += '-DSYSTEM_VIEW '
```
Comment (prusnak) [inline core/SConscript.firmware:367]: Isn't this redundant because of the row above doing the same?

File: core/embed/firmware/mphalport.c [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
 #include "common.h"
 #include "py/mphal.h"
+#include "string.h"
```
Comment (prusnak) [inline core/embed/firmware/mphalport.c:22]: Why?

File: core/embed/firmware/systemview.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+} SCS_REGS;
+
+extern uint32_t SystemCoreClock;
+
+static inline uint32_t is_mode_unprivileged(void) {
+  uint32_t r0;
+  __asm__ volatile("mrs %0, control" : "=r"(r0));
+  return r0 & 1;
+}
+
```
Comment (prusnak) [inline core/embed/firmware/systemview.c:58]: Can be replaced with `#include "supervise.h"`

**Review by prusnak — CHANGES_REQUESTED** [review]

File: core/SConscript.firmware (hunk 1 above)
Comment (hiviah) [reply core/SConscript.firmware:367]: No, each one affects different invocation of gcc, there is one for firmware and other for micropython.

File: core/embed/firmware/mphalport.c (hunk 2 above)
Comment (hiviah) [reply core/embed/firmware/mphalport.c]: Forgotten from previous implementations. Removed.

File: core/embed/firmware/systemview.c (hunk 3 above)
Comment (hiviah) [reply core/embed/firmware/systemview.c]: supervise.h wasn't enough and there used to be linker conflict with duplicate symbols. Found some combination of includes that fixes it.

**Pull request merged**